File: services/rigor/redline/correction/bundle.py
"""Assemble the downloadable correction bundle: a README and a consolidated notebook.

Given one row per finding, `build_readme` writes what was wrong, what each script fixes,
and how to run them, and `build_notebook` writes a single .ipynb (nbformat 4) with a
header cell plus one markdown and one code cell per finding.

The honesty rule holds here too. An unsalvageable finding gets its markdown cell saying
the design cannot be rescued from this data and what would be needed instead, and it
carries no code cell claiming a fix. There is no path that renders a corrected result for
a finding with none.

stdlib only: this module builds JSON with the standard library so it imports without the
heavy stack.
"""

from __future__ import annotations

import json
from typing import Any, Sequence

Finding = dict[str, Any]


def _title(row: Finding) -> str:
    cid = row.get("checkId", "?")
    name = str(row.get("name", "finding"))
    return f"Check {cid}: {name}"


def _fix_line(row: Finding) -> str:
    """One sentence on what this script does, honest about the unsalvageable case."""
    if row.get("unsalvageable"):
        return (
            "This design cannot be rescued from this data, so there is no corrected "
            "result. The script prints the verdict and what new data would be needed."
        )
    corrected = row.get("corrected")
    if corrected is None:
        return "This script re-runs the analysis the honest way and prints the result."
    return (
        "This script re-runs the analysis the honest way. The corrected result is "
        f"{corrected}, printed as the last line so you can check it against the report."
    )


def build_readme(findings: Sequence[Finding]) -> str:
    """A short README: what was wrong, what each script fixes, how to run them."""
    lines: list[str] = []
    lines.append("# Redline corrected analysis")
    lines.append("")
    lines.append(
        "Redline re-ran the load-bearing statistics in your analysis. Each script below "
        "reproduces one honest re-analysis. Every script is self-contained and takes the "
        "path to your data:"
    )
    lines.append("")
    lines.append("```")
    lines.append("python <script>.py --h5ad your_data.h5ad")
    lines.append("```")
    lines.append("")
    lines.append(
        "Each script prints a human-readable report and, as its last line, a "
        "`REDLINE_RESULT` JSON object. That line is the machine-readable result, and it "
        "matches the numbers in the Redline report by construction."
    )
    lines.append("")
    if not findings:
        lines.append("No findings were flagged. Every check reported clean.")
        return "\n".join(lines) + "\n"

    lines.append("## Findings")
    lines.append("")
    for row in findings:
        lines.append(f"### {_title(row)}")
        lines.append("")
        filename = str(row.get("filename", ""))
        if row.get("error"):
            lines.append(f"Could not complete: {row['error']}")
            lines.append("")
            continue
        lines.append(_fix_line(row))
        lines.append("")
        if filename and not row.get("unsalvageable"):
            lines.append("```")
            lines.append(f"python {filename} --h5ad your_data.h5ad")
            lines.append("```")
            lines.append("")
    return "\n".join(lines).rstrip() + "\n"


def _md_cell(source: str) -> dict[str, Any]:
    return {"cell_type": "markdown", "metadata": {}, "source": source.splitlines(keepends=True)}


def _code_cell(source: str) -> dict[str, Any]:
    return {
        "cell_type": "code",
        "execution_count": None,
        "metadata": {},
        "outputs": [],
        "source": source.splitlines(keepends=True),
    }


def build_notebook(findings: Sequence[Finding]) -> str:
    """A consolidated notebook (serialized .ipynb JSON, nbformat 4).

    One header cell, then per finding a markdown cell and, when a fix exists, a code cell.
    An unsalvageable finding gets its markdown cell and no code cell.
    """
    cells: list[dict[str, Any]] = []
    header = [
        "# Redline corrected analysis\n",
        "\n",
        "This notebook collects the honest re-analyses for every flagged finding. ",
        "Each code cell is a self-contained script. Point it at your data by setting ",
        "the `--h5ad` path, or run it from a shell.\n",
        "\n",
        "The last printed line of each script is a `REDLINE_RESULT` JSON object, the ",
        "machine-readable result that matches the Redline report.\n",
    ]
    cells.append({"cell_type": "markdown", "metadata": {}, "source": header})

    for row in findings:
        title = _title(row)
        filename = str(row.get("filename", ""))
        if row.get("error"):
            cells.append(_md_cell(f"## {title}\n\nCould not complete: {row['error']}\n"))
            continue
        if row.get("unsalvageable"):
            md = (
                f"## {title}\n\n"
                "This design cannot be rescued from this data. There is no corrected "
                "result to show. To answer the question, the experiment needs a design "
                "that separates the effect of interest from the confound, which means new "
                "data, not a re-analysis of this data.\n"
            )
            cells.append(_md_cell(md))
            continue
        md = f"## {title}\n\n{_fix_line(row)}\n\nFile: `{filename}`\n"
        cells.append(_md_cell(md))
        code = str(row.get("code", ""))
        if code:
            cells.append(_code_cell(code))

    notebook = {
        "cells": cells,
        "metadata": {
            "kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"},
            "language_info": {"name": "python"},
        },
        "nbformat": 4,
        "nbformat_minor": 5,
    }
    return json.dumps(notebook, indent=1)


__all__ = ["build_notebook", "build_readme"]

File: packages/engine/src/bundle.ts
import type {
  AuditReport,
  CheckResult,
  CorrectedBundle,
  CorrectedScript,
  DatasetMeta,
  Recommendation,
} from '@redline/contracts';
import { checkMeta } from '@redline/contracts';

/**
 * The corrected-analysis bundle: the artifact that outlasts the demo week. It
 * collects, for every finding that has a fix, the runnable script, a README, and
 * a consolidated notebook.
 *
 * The honesty rule holds here too. Clean findings are skipped (nothing to
 * correct). An unsalvageable finding gets a notebook cell that says the design
 * cannot be rescued from this data and what would be needed, and NO code cell
 * claiming a fix. Its non-separability proof script is still included, because
 * proving the dead end is honest; inventing a corrected number is not.
 */
export function buildBundle(report: AuditReport, dataset: DatasetMeta): CorrectedBundle {
  const findings = report.results.filter((r) => r.state !== 'clean');
  const scripts = collectScripts(findings);
  return {
    readme: buildReadme(report, dataset, findings),
    notebook: buildNotebook(report, dataset, findings),
    scripts,
  };
}

function isUnsalvageable(r: CheckResult): boolean {
  return r.preview?.unsalvageable === true;
}

/** One script per finding that carries corrected code (flagged findings do). */
function collectScripts(findings: CheckResult[]): CorrectedScript[] {
  const out: CorrectedScript[] = [];
  for (const r of findings) {
    if (!r.correctedCode) continue;
    out.push({
      checkId: r.checkId,
      title: checkMeta(r.checkId).name,
      filename: r.correctedCode.filename,
      code: r.correctedCode.inline,
    });
  }
  return out;
}

function firstNeedsData(recs: Recommendation[] | undefined): Recommendation | undefined {
  return recs?.find((x) => x.feasibility === 'needs_new_data' || x.feasibility === 'unsalvageable');
}

function buildReadme(report: AuditReport, dataset: DatasetMeta, findings: CheckResult[]): string {
  const lines: string[] = [];
  lines.push('# Redline corrected analysis');
  lines.push('');
  lines.push(`Dataset: ${dataset.title}`);
  lines.push(`File: ${dataset.file}`);
  lines.push('');
  lines.push(report.verdict);
  lines.push('');
  if (findings.length === 0) {
    lines.push('Every check passed clean. There is nothing to correct.');
    return lines.join('\n');
  }
  lines.push('Each script below re-runs one finding the honest way. Every script');
  lines.push('takes --h5ad PATH and prints its result as a REDLINE_RESULT line, so');
  lines.push('the numbers are reproducible from your own data.');
  lines.push('');
  for (const r of findings) {
    const meta = checkMeta(r.checkId);
    lines.push(`## Check ${r.checkId}: ${meta.name}`);
    if (r.error) lines.push(`Problem: ${r.error}.`);
    if (isUnsalvageable(r)) {
      lines.push('This finding is unsalvageable. There is no valid fix on this data,');
      lines.push('so no corrected result is provided.');
      const needs = firstNeedsData(r.recommendations);
      if (needs) lines.push(`What would be needed: ${needs.action}`);
    } else if (r.correctedCode) {
      lines.push(`Corrected: ${r.corrected}`);
      lines.push(`Run: ${r.correctedCode.entrypoint}`);
    }
    lines.push('');
  }
  return lines.join('\n').trimEnd() + '\n';
}

// ── Notebook (nbformat 4) ────────────────────────────────────────────────────

interface NbCell {
  cell_type: 'markdown' | 'code';
  metadata: Record<string, never>;
  source: string[];
  execution_count?: null;
  outputs?: never[];
}

function toSource(text: string): string[] {
  const lines = text.split('\n');
  return lines.map((l, i) => (i < lines.length - 1 ? l + '\n' : l));
}

function mdCell(text: string): NbCell {
  return { cell_type: 'markdown', metadata: {}, source: toSource(text) };
}

function codeCell(text: string): NbCell {
  return { cell_type: 'code', metadata: {}, execution_count: null, outputs: [], source: toSource(text) };
}

function findingMarkdown(r: CheckResult): string {
  const meta = checkMeta(r.checkId);
  const parts: string[] = [`## Check ${r.checkId}: ${meta.name}`, ''];
  if (r.error) parts.push(`Problem: ${r.error}.`, '');
  if (isUnsalvageable(r)) {
    parts.push('This design cannot be rescued from this data. There is no valid fix,');
    parts.push('so this notebook carries no corrected code for it.');
    const needs = firstNeedsData(r.recommendations);
    if (needs) parts.push('', `What would be needed: ${needs.action}`);
  } else {
    parts.push(`Corrected: ${r.corrected}`);
  }
  return parts.join('\n');
}

function buildNotebook(report: AuditReport, dataset: DatasetMeta, findings: CheckResult[]): string {
  const cells: NbCell[] = [
    mdCell(
      [
        '# Redline corrected analysis',
        '',
        `Dataset: ${dataset.title} (${dataset.file})`,
        '',
        report.verdict,
      ].join('\n'),
    ),
  ];
  for (const r of findings) {
    cells.push(mdCell(findingMarkdown(r)));
    // No code cell for an unsalvageable finding: there is no fix to run.
    if (!isUnsalvageable(r) && r.correctedCode) {
      cells.push(codeCell(r.correctedCode.inline));
    }
  }
  const nb = {
    cells,
    metadata: {
      kernelspec: { display_name: 'Python 3', language: 'python', name: 'python3' },
      language_info: { name: 'python' },
    },
    nbformat: 4,
    // nbformat_minor 4, not 5: the 4.5 schema makes a per-cell `id` mandatory,
    // and these cells carry none, so a 4.5 notebook fails nbformat.validate with
    // one "'id' is a required property" per cell (a future nbformat makes that a
    // hard error). The cells use no 4.5-only feature, so 4.4 is the honest,
    // valid version to declare.
    nbformat_minor: 4,
  };
  return JSON.stringify(nb, null, 1);
}
